README.org
* childsds
** List of available references/LMS charts
  - you find a list of available reference tables by region and by item [[https://github.com/mvogel78/childsds/wiki][here]].
** Installation 
   - version 0.6.4 from cran via ~install.packages()~ or the respective menu
** Basic use
*** Transformation into sds
#+BEGIN_SRC R :session :exports both :results output
  library(childsds)
  ## generate example data
  x <- data.frame(height=c(50,100,60,54),
		  sex=c("m","f","f","m"),
		  age=c(0,2.9,0.6,0.2))
  x$height.sds <- sds(value = x$height,
		      age = x$age,
		      sex = x$sex, male = "m", female = "f",
		      ref = who.ref, item = "height")
  head(x)
#+END_SRC

#+RESULTS[46a6ba3828dbb6c977bc976a6280e0b191bc02ee]:
:   height sex age  height.sds
: 1     50   m 0.0  0.06116878
: 2    100   f 2.9  1.54150151
: 3     60   f 0.6 -3.26293906
: 4     54   m 0.2 -2.82189275

*** make_percentile_tab()
**** create a percentile table
#+BEGIN_SRC R :session :exports both :results output
  library(childsds)
  head(tab <- make_percentile_tab(ref = nl4.ref,
				  item = "heightM",
				  perc = c(5,50,95),
				  age = 1:3))
#+END_SRC

#+RESULTS:
:      sex age perc_05_0 perc_50_0 perc_95_0 nu       mu      sigma
: 1   male   1  72.82291  77.15261  81.48232  1 77.15261 0.03411775
: 2   male   2  82.10371  87.67000  93.23629  1 87.67000 0.03860000
: 3   male   3  89.97701  96.28000 102.58299  1 96.28000 0.03980000
: 4 female   1  70.58366  74.89305  79.20245  1 74.89305 0.03498225
: 5 female   2  82.06492  86.76000  91.45508  1 86.76000 0.03290000
: 6 female   3  89.41744  94.83000 100.24256  1 94.83000 0.03470000

**** use the ~stack~ argument to create a dataframe in the long format for use in ggplot 
#+BEGIN_SRC R :session :exports both :results output
  library(childsds)
  head(tab <- make_percentile_tab(ref = nl4.ref,
				  item = "heightM",
				  perc = c(5,50,95),
				  age = seq(0,20,by=0.1),
				  stack = T))
#+END_SRC

#+RESULTS:
:   age  sex  variable    value
: 1 0.0 male perc_05_0 47.82905
: 2 0.1 male perc_05_0 51.65139
: 3 0.2 male perc_05_0 55.37913
: 4 0.3 male perc_05_0 58.68443
: 5 0.4 male perc_05_0 61.60275
: 6 0.5 male perc_05_0 64.21947

#+BEGIN_SRC R :session :exports both :results output graphics :file fig_1.png :height 400 :width 600
  library(ggplot2)
  ggplot(tab, aes( x = age, y = value, group=paste(sex, variable))) +
      geom_line(aes(linetype = sex)) +
      theme_classic() +
      theme(legend.position = c(0.1,0.8))
#+END_SRC

#+RESULTS[9e68fab1cedee6b9007e7fe1696cac77f23d3ef4]:
[[file:fig_1.png]]


NEWS.md
# childsds 0.6.7
## Major changes
* add uk1990 references
* add liver enzymes references
## Minor changes
# childsds 0.6.6
## Major changes
* add ggplot version of the worm plot function
## Minor changes
* add include.pars to make_percentile_tab
* add transformation argument to lms() wrapper
# childsds 0.6.5
## Major changes
* add who 2007 reference tables
* add nl3 reference tables
* add preterm reference tables
* add turkish reference tables height, weight
* add igf international and japan
* add saudi arabian preschool references
## Bug fixes
* removes class labelled from who references
* add missing argument to one_it inside do_iterations
# childsds 0.6.4
## Major changes
* add colombian skinfold references
## Bug fixes
*  fix typos in italian references
# childsds 0.6.3
## Major changes
* add option whether or not models should be kept during the iteration process
## Minor changes
* add subscapular skinfold to kiggs.ref
* add function to find better upper and lower age bounds (full month)
## Bug fixes
*  remove user accessable documentation of unexported functions

# childsds 0.6.2
## Major changes
* add sds_2d for calculation of bloodpressure sds
* add kiggs bloodpressure references
* add additional new references (Belgium)

## Bug fixes
* Remove direct calls to dplyr::n()

# childsds 0.6.1

* Added a `NEWS.md` file to track changes to the package.

## Major changes
* add new references
* add function make_percentile_tab() to extract data for use in ggplot()
## Bug fixes
* fix namespace problems for gamlss distributions
* fix missing export of sds() function



R/sdsvals.r
##' Calculate SDS values
##'
##' The function takes a vector of measurement values, and of age and of sex
##' and a RefGroup object as arguments. It calculates the sds or percentile
##' values.
##' @title Calculate SDS Values
##' @param value vector of measurement values
##' @param age vector of age values
##' @param sex vector of sex 
##' @param item name of the item e.g. "height"
##' @param ref RefGroup object
##' @param type "SDS" or "perc"
##' @param male coding of sex for male
##' @param female coding of sex for female
##' @return vector containing SDS or percentile values
##' @examples
##' anthro <- data.frame(age = c(11.61,12.49,9.5,10.42,8.42,10.75,9.57,10.48),
##'                      height = c(148.2,154.4,141.6,145.3,146,140.9,145.5,150),
##'                      sex = sample(c("male","female"), size = 8, replace = TRUE),
##'                      weight = c(69.5,72.65,47.3,51.6,45.6,48.9,53.5,58.5))
##' anthro$height_sds <- sds(anthro$height,
##'                          age = anthro$age,
##'                          sex = anthro$sex, male = "male", female = "female",
##'                          ref = kro.ref,
##'                          item = "height",
##'                          type = "SDS")
##' 
##' anthro$bmi <- anthro$weight/(anthro$height**2) * 10000
##' anthro$bmi_perc <- sds(anthro$bmi,
##'                        age = anthro$age,
##'                        sex = anthro$sex, male = "male", female = "female",
##'                        ref = kro.ref,
##'                        item = "bmi",
##'                        type = "perc")
##' data(who.ref)
##' x <- data.frame(height=c(50,100,60,54),
##'                 sex=c("m","f","f","m"),
##'                 age=c(0,2.9,0.6,0.2))
##' sds(value = x$height, age = x$age, sex = x$sex, male = "m", female = "f",
##'     ref = who.ref, item = "height")
##' @author Mandy Vogel
##' @export
sds <- function(value, age, sex, item, ref, type = "SDS", male = "male", female = "female"){
    if(!(length(value) == length(age) & length(value) == length(sex))){
        print("value, age, and sex must be of the same length")
        invisible(return(NULL))
    }
    if(!item %in% names(ref@refs)){
        stop(paste(item,"is not available in the given refs object.",
                    "Please choose one of the following items:",
                    paste(names(ref@refs), collapse = " - ")))
        return(invisible(NULL))
    }
    sex <- as.character(factor(sex, levels = c(male, female), labels = c("male", "female")))
    refs <- ref@refs[[item]]@params
    dists <- ref@refs[[item]]@dist
    par.appr <- lapply(refs, function(df){
        as.data.frame(lapply(df, function(param) stats::approx(df$age, param, xout = age, rule = 1)$y))})
    res <- numeric()
    for(i in 1:length(value)) {
        if(is.na(value[i]) | any(is.na(unlist(par.appr[[sex[i]]][i,-1]))) ){
            res[i] <- NA
        } else{
        res[i] <- eval(parse(text = paste0("gamlss.dist::p",dists[[sex[i]]],"(",value[i],",",
                                           paste(paste(names(par.appr[[sex[i]]])[-1],"=", par.appr[[sex[i]]][i,-1]), collapse = ","),
                                           ")")))
        }
    }
    if(type == "SDS") return(stats::qnorm(res))
    return(round(res * 100, 2))
}


##' Calculate SDS values for 2-dimensional matrix of covariates 
##'
##' The function takes a vector of measurement values, and of age and a
##' second covariate (like age and height for blood pressure) of sex
##' and a RefGroup object as arguments. It calculates the sds or percentile
##' values. This function is beta.
##' @title Calculate SDS Values for 2-dimensional matrix of covariates
##' @param value vector of measurement values
##' @param age vector of age values
##' @param x2 second vector of covariates
##' @param sex vector of sex 
##' @param item name of the item e.g. "height"
##' @param ref RefGroup object
##' @param type "SDS" or "perc"
##' @param male coding of sex for male
##' @param female coding of sex for male
##' @details the function searches for the nearest given point in the reference grid.
##' From there, the SDS/percentile value will be calculated. Different from \code{\link{sds}},
##' no interpolation will be applied. The procedure is according to Neuhauser et al. Blood
##' Pressure Percentiles by Age and Height from Nonoverweight Children and Adolescents
##' in Germany. 2011.
##' @return vector containing SDS or percentile values
##' @author Mandy Vogel
##' @export
sds_2d <- function(value, age, x2, sex, item, ref, type = "SDS", male = "male", female = "female"){
    if(!(length(value) == length(age) &
         length(value) == length(sex) &
         length(value) == length(x2))){
        print("value, age, x2, and sex must be of the same length")
        invisible(return(NULL))
    }
    if(!item %in% names(ref@refs)){
        stop(paste(item,"is not available in the given refs object.",
                    "Please choose one of the following items:",
                    paste(names(ref@refs), collapse = " - ")))
        return(invisible(NULL))
    }
    sex <- as.character(factor(sex,
                               levels = c(male, female),
                               labels = c("male", "female")))
    refs <- ref@refs[[item]]@params
    dists <- ref@refs[[item]]@dist
    par.appr <- list()
    par.appr <- lapply(refs, function(df){
        tmpdf <- data.frame(
                   mu = df$mu[class::knn(df[,c("age","x2")],
                                         data.frame(age = age,
                                                    x2 = x2),
                                         k = 1,cl = 1:nrow(df))],
                   nu = df$nu[class::knn(df[,c("age","x2")],
                                         data.frame(age = age,
                                                    x2 = x2),
                                         k = 1,cl = 1:nrow(df))],
                   sigma = df$sigma[class::knn(df[,c("age","x2")],
                                               data.frame(age = age,
                                                          x2 = x2),
                                               k = 1,cl = 1:nrow(df))]
                   ) 
        tmpdf
    })
    res <- numeric()
    for(i in 1:length(value)) {
        if(is.na(value[i]) | any(is.na(unlist(par.appr[[sex[i]]][i,]))) ){
            res[i] <- NA
        } else{
            print(paste0("gamlss.dist::p",dists[[sex[i]]],"(",value[i],",",
                                           paste(paste(names(par.appr[[sex[i]]]),"=", par.appr[[sex[i]]][i,]), collapse = ","),
                                           ")"))
        res[i] <- eval(parse(text = paste0("gamlss.dist::p",dists[[sex[i]]],"(",value[i],",",
                                           paste(paste(names(par.appr[[sex[i]]]),"=", par.appr[[sex[i]]][i,]), collapse = ","),
                                           ")")))
        }
    }
    if(type == "SDS") return(stats::qnorm(res))
    return(round(res * 100, 2)) 
}
